docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,1171 @@
+>
+>   * A TEE provider (like SGX) is responsible for implementing a plug-in that knows how seal/unseal on that TEE platform and _defines a plug-in-specific cryptosuite and data blob format._
+>
+>     * This shifts the crypto-agility problem out of the SDK; a developer can adopt alternative implementations of sealing plugins or plugins using different (newer) cryptosuites without relying on the SDK to determine that on their behalf.
+>   * OE SDK provides `oe_seal` and `oe_unseal` APIs that are wrappers over TEE plug-in implementation.
+>
+>     * Many of the ideas from attestation plug-ins may carry over here, such as UUID targeting, plug-in registration, etc. Some will get simplified (e.g. UUID is always embedded in sealed blob because there is no cross-TEE sealing)
+>     * Some or all of the OE semantics may go away attempting to generalize sealing policies across TEEs, much in the same way that attestation plugins don't attempt to build on top of the old `oe_parse_report` semantics.
+>     * For example, consider a trivial use case where the user has no control over sealing parameters at all if they write cross-TEE code; `oe_seal` just calls the appropriate TEE-plugin and it seals with a default policy of its choice (e.g. always with MRENCLAVE for SGX). The TEE-plugin may then expose its custom set of properties through an opaque blob type through the `oe_seal` API that a user that wants to customize functionality on a specific TEE can manipulate themselves (e.g. `sgx_key_request` struct for SGX).
+>
+> I'm not proposing that we actually run off and go build this strawman (there are some other problems I'm glossing over here like the `oe_get_private_key` API) but it represents a high level concept for an alternative that I think hits some of the stated objectives better:
+>
+> * Actually TEE-agnostic, bakes less implementation directly into OE SDK (or at least, could be more easily broken out as standalone components moving forwards)
+> * Accommodative in that the plug-ins get to expose the full range of options specific to a TEE if desired without it filtering through an OE SDK aggregation layer.
+>
+>   * The trade-off being less expressive sealing semantics for apps that want to span different TEEs, at least in this strawman version.
+> * Easy to use in that there's no management of key info at all, and that's incorporated into the `oe_seal` function.
+>
+>   * With proper built-in defaults per TEE, it could also avoid some of the registration complexity associated with attestation plugins, again because there's a much stronger binding of TEE the enclave is running in vs. attestation.
+> * Interoperable with existing Intel SDK, since Intel controls the plugin entirely, they can choose to provide that functionality as they see fit.
+
+Plug-ins allows multiple implementations to coexist and allows the developer to choose which plug-ins to go into the enclave. Based on the analyses in previous sections, plug-ins are not necessary IMHO.
```

Comment:
I'll provide a different opinion here which is that the goal of thinking about sealing implementations as plug-ins is more than just allowing multiple implementations to coexist at runtime, it is about defining a consistent set of expectations for how an arbitrary implementation interacts with the OE SDK and other plug-ins. To have something concrete to discuss, consider a formal plugin framework that builds on the alternative public `oe_seal()` proposal above:

```c
/**
 * Extend the types of settings supported by oe_seal to
 * include plug-in specific settings
 */
typedef enum _oe_seal_setting_type
{
    OE_SEAL_SETTING_TYPE_AGNOSTIC = 0xd764778b,
    OE_SEAL_SETTING_TYPE_SGX = 0x8ada353c,
    OE_SEAL_SETTING_TYPE_PLUGIN = 0x887d78bc
} oe_seal_setting_type_t;

typedef struct _oe_seal_setting_plugin
{
    oe_seal_setting_header_t header;
    oe_uuid_t plugin_id;
    const void* custom_data;
    size_t custom_data_size;
} oe_seal_setting_plugin_t;

#define OE_INIT_SEAL_SETTING_PLUGIN(PLUGIN_ID, CUSTOM_DATA, CUSTOM_DATA_SIZE) \
    {                                                                         \
        .header.type = OE_SEAL_SETTING_TYPE_PLUGIN,                           \
        .header.size = sizeof(oe_seal_setting_plugin_t),                      \
        .plugin_id = PLUGIN_ID, .custom_data = CUSTOM_DATA,                   \
        .custom_data_size = CUSTOM_DATA_SIZE                                  \
    }
```

For a basic plug-in model, we would also need some management functions. Note that the rest of this discussion considers these to also be in the public headers, but are primarily of interest to sealing implementation providers:

```c
/**
 * Definition for registering a seal plugin with the OE SDK
 */
typedef struct _oe_seal_plugin_definition
{
    /* UUID for the seal plugin */
    oe_uuid_t plugin_id;

    /* Callback implemented by the plugin to handle oe_seal().
     * Note that seal_callback expands the function prototype of
     * oe_seal() to pass plugin-defined custom_data to the plugin */
    oe_result_t (*seal_callback)(
        const oe_seal_setting_t* settings,
        size_t settings_count,
        const void* custom_data,
        size_t custom_data_size,
        const uint8_t* plaintext,
        size_t plaintext_size,
        const uint8_t* additional_data,
        size_t additional_data_size,
        uint8_t** blob,
        size_t* blob_size);

    /* Callback implemented by the plugin to handle oe_unseal().
     * This has the same function prototype as oe_unseal(). */
    oe_result_t (*unseal_callback)(
        uint8_t* blob,
        size_t blob_size,
        uint8_t** plaintext,
        size_t* plaintext_size,
        uint8_t** additional_data,
        size_t* additional_data_size);
} oe_seal_plugin_definition_t;

/**
 * Register a plugin for oe_seal() and oe_unseal().
 *
 * @param[in] plugin Definition of the plugin to register.
 * @param[in] make_default Sets the plugin to be the default for `oe_seal()`
 * and `oe_unseal()`
 *
 * @retval OE_OK Plugin was successfully registered or already registered.
 * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
 * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
 */
oe_result_t oe_register_seal_plugin(
    oe_seal_plugin_definition_t plugin,
    bool make_default);

/**
 * Unregister a specific sealing plug-in
 *
 * @param[in] plugin_id UUID of the plugin to unregister.
 *
 * @retval OE_OK Plugin was successfully unregistered.
 * @retval OE_NOT_FOUND Specified plugin is not registered.
 */
oe_result_t oe_unregister_seal_plugin(oe_uuid_t plugin_id);
```

As an example, consider a new SGX sealing plugin that wants to allow the developer to specify a label/context value for the plugin to
use as part of its key derivation, instead of reserving that field for its own use as with the existing SGX sealing implementation.
To participate in the plugin framework, it would need provide:

```c
/**
 * Public UUID for the new plugin
 */
#define OE_SEAL_PLUGIN_UUID_CUSTOM_SGX                                    \
    {                                                                     \
        0xd2, 0x5f, 0x43, 0x09, 0x4f, 0x66, 0x40, 0x31, 0xa2, 0xab, 0xf1, \
            0x97, 0x3c, 0xbf, 0x8c, 0xbe                                  \
    }

/**
 * Implementation of its seal_callback
 */ 
oe_result_t sgx_custom_seal(
    const oe_seal_setting_t* settings,
    size_t settings_count,
    const void* custom_data,
    size_t custom_data_size,
    const uint8_t* plaintext,
    size_t plaintext_size,
    const uint8_t* additional_data,
    size_t additional_data_size,
    uint8_t** blob,
    size_t* blob_size)
{
    ...
}

/**
 * Implementation of its unseal_callback
 */ 
oe_result_t sgx_custom_unseal(
    uint8_t* blob,
    size_t blob_size,
    uint8_t** plaintext,
    size_t* plaintext_size,
    uint8_t** additional_data,
    size_t* additional_data_size)
{
    ...
}

/**
 * Public definition of the plugin for the developer to register
 */
extern const oe_seal_plugin_definition_t OE_SEAL_PLUGIN_CUSTOM_SGX = {
    .plugin_id = OE_SEAL_PLUGIN_UUID_CUSTOM_SGX,
    .seal_callback = sgx_custom_seal,
    .unseal_callback = sgx_custom_unseal
};

```

Consider the example use case driving the need to support coexisting sealing implementations, where a developer wants to migrate sealed data from one implementation to another. For this example, let's say it is to migrate data sealed using the default OE SDK implementation to the custom SGX sealing plugin (and for the sake of readability, we will elide the result checking from the function calls):

```c
// Register the plugin definition
oe_result_t result = oe_register_seal_plugin(OE_SEAL_PLUGIN_CUSTOM_SGX, false);

// Unseal the data that was previously sealed by the enclave using the default sealer.
result = oe_unseal(
    ciphertext,
    ciphertext_size,
    &decrypted_text,
    &decrypted_text_size,
    &authenticated_additional_data,
    &authenticated_additional_data_size);

// Define custom configuration as specified by the plugin implementation
char* custom_config =
    "{ \"label\": \"Value to hash into KEYREQUEST.KEYID structure (32-bytes)\" }";
size_t custom_config_size = (size_t)strlen(custom_config) + 1;

// Configure the seal settings array to specify the custom SGX seal plugin
sgx_attributes_t attributes_mask = { 0xFFFFFFFF, 0 };
my_seal_settings_t seal_settings = {
    .sgx_settings = OE_INIT_SEAL_SETTING_SGX(KEYPOLICY_MRENCLAVE, attributes_mask, 0),
    .plugin_settings = OE_INIT_SEAL_SETTING_PLUGIN(
        OE_SEAL_PLUGIN_UUID_CUSTOM_SGX, custom_config, custom_config_size)};

// Seal the data using the custom SGX seal plugin
result = oe_seal(
    (oe_seal_setting_t*)&seal_settings,
    2,
    my_plaintext,
    my_plaintext_size,
    my_additional_data,
    my_additional_data_size,
    &ciphertext,
    &ciphertext_size);

```

It's not an incredible amount of overhead to support this scenario, and it asserts a specific distribution of responsibilities between the components that we can discuss in greater detail:

#### OE SDK requirements

* OE SDK must provide a default sealing implementation for each TEE platform that supports sealing.
  * The default implementation is itself implemented as a plugin and adheres to the same set of expectations as other plugins.
  * The default implementation can be registered via global constructor defined by the OE SDK itself.
  * The default implementation must support all OE-defined seal settings (i.e. `oe_seal_setting_agnostic_t` and `oe_seal_setting_sgx_t` for the default SGX sealer).
* OE SDK tracks a global list of registered sealing plugin definitions for `oe_register_seal_plugin()`/`oe_unregister_seal_plugin()`.
  * As an implementation suggestion, the list may be treated as ordered, with the head definition being the default.
    * Registering with `make_default` effectively inserts the definition at the head instead of the tail of the list.
    * Removing a plugin marked as default will make the next definition default.
    * Removing all plugins or all usable plugins for the current TEE platform will result in `oe_seal()`/`oe_unseal()` returning `OE_NOT_IMPLEMENTED`.
    * Access to the list is guarded by mutex for thread safety.
* `oe_seal()` must scan the list of provided `oe_seal_settings` for any specified plugins.
  * If any plugins are specified, they are invoked in the order in which they are specified and before the default plugin for that platform (if different from the specified plugins).
    * Note that this proposal does not require plugins to be specific to a given TEE; a plugin may be itself implemented to be compatible on different TEEs.
* `oe_unseal()` must invoke all registered plugins beginning with the default, until the `oe_unseal()` operation is handle.
  * This can be optimized a bit by requiring a standard data contract for all sealing plugins to use an OE-defined header that specifies the `plugin_id` of the implementation that produced it.
    * The tradeoff is that OE SDK will not support direct compatibility with sealed blobs produced by other SDKs (e.g. Intel SGX SDK).
    * This is likely acceptable because hardware-sealed blobs are generally not intended to be unsealable on a different device, and may not be unsealable by a different implementation of the enclave depending on the sealing key policy.
      * i.e. the only existing use case is very narrow: migrating data sealed by an Intel SGX SDK enclave using the MRSIGNER policy to an OE SDK enclave _on the same machine_.
    * If a standard sealed blob header is used, OE SDK takes responsibility for parsing the standard header and invoking the appropriate plugin implementation by its UUID. Otherwise, each plugin is invoked in turn and is responsible for its own format probing logic until one plugin accepts the blob.

#### Sealing plugin requirements
* `seal_callback` should be able to function without additional parameters.
  * This includes choosing a good default key policy, ideally consistent with OE SDK's default implementation (e.g. if the default implementation uses `OE_SEAL_POLICY_PRODUCT`, we would encourage plugins to use the closest equivalent default).
* `seal_callback` should use information from `oe_seal_setting_t` where appropriate.
  * It is passed all of the `oe_seal_setting_t` from `oe_seal()` and it must ignore (not error) on any settings it does not support.
    * This is necessary for a developer to write their sealing code once, provide all variants of the custom sealing parameters that may exist on one TEE platform or another.
  * This is preferred over `seal_callback` taking all its parameters from `custom_data` although nothing in the framework prevents this.
  * The plugin has the option to privilege more specific options where they conflict with more general `oe_seal_settings_t`
    * For example, a caller may provide the agnostic `OE_SEAL_POLICY_PRODUCT` setting and also a more specific SGX `KEYPOLICY_CONFIGID | KEYPOLICY_ENCLAVE` setting, and the more specific SGX setting should apply.
* `seal_callback` is solely responsible for defining, versioning and interpreting any `custom_data` it accepts.
  * The parameter is entirely opaque to the OE SDK and is passed directly to the sealing implementation.
  * OE SDK may recommend that the `custom_data` be structured in a standard way (e.g. as JSON in the example), but it provides no enforcement. `custom_data` is entirely a private contract between the caller and the sealing plugin that is not mediated by OE SDK.
* If a plugin is executed in an unsupported environment, it can return `OE_UNSUPPORTED`, and OE SDK will attempt to invoke the next specified plugin (or default for the TEE, if different).
* All plugins must support some form of authenticated additional data for the `additional_data` parameter.
  * This is not a new requirement, just calling out the existing requirement reflected in the original `oe_seal()` function prototype.
    * If we relax this requirement, a sealing plugin should at least be expected to error our with `OE_UNSUPPORTED` instead of silently dropping the `additional_data` parameter (or any standard parameter on the `oe_seal()` API).
  * Note that this proposal places no other constraints on the choice of cryptography used by the plugin. It does not suggest key lengths, key derivation flows, encryption or authenticated hashing algorithms.

Reply:
@CodeMonkeyLeet, I have the same question as above: What are the benefits that could be brought in by this plugin model? You mentioned that

> it is about defining a consistent set of expectations for how an arbitrary implementation interacts with the OE SDK and other plug-ins. 

But I'm not sure what you meant by "arbitrary implementation". In practice, it's always the sealer that specifies the plugin. And given the unsealer is the same entity (i.e., the same enclave as the sealer, or signed by the same key as the sealer), it inherently knows which plugin to use. I don't see why a plugin needs to interact with any entities other than itself. Wouldn't it be simpler to link to the plugin directly, than to look up the plugin using UUID at runtime?

Reply:
@binxing By arbitrary implementation, I mean that sealing functionality can be implemented and consumed by OE SDK users without inclusion or review by OE SDK itself. Plugin models create a consistent set of guidelines and expectations for the user, 

By definition, the plugin already interacts with the OE SDK, and that doesn't change with the introduction of a formal plugin framework (otherwise, it's just a library and we wouldn't be having a conversation about adding/changing public OE SDK APIs at all). This simply formalizes the relationship between OE SDK and a plugin implementation, so that OE SDK can ensure a consistent user experience for how it interacts with the plugins on the user's behalf (including what it does and does not mediate). 

Regarding whether it's simpler to link the plugin directly, the question is simpler for whom and in what context?
- My priorities in thinking about the public API would be that the semantics for the OE SDK user are clear and easy to use, and I'm willing to trade off having the OE SDK & sealing implementation wrap a bit more work.
- In the case where we require a developer to link separate libraries and then invoke underlying methods provided by different libraries as needed, if you have more than one implementation, then the `oe_seal()` weak alias implementation is determined by link order and that can violate the Principle of Least Astonishment when a change in build definitions create a change in runtime behavior. It is not simpler in that regard.

Reply:
> By arbitrary implementation, I mean that sealing functionality can be implemented and consumed by OE SDK users without inclusion or review by OE SDK itself. Plugin models create a consistent set of guidelines and expectations for the user,

I agree with you that it's important to have consistent guidelines and expectations. But why does that require plugin models? `oe_seal()` and `oe_unseal()` themselves already set forth guidelines and expectations. After all, the public APIs do nothing except dispatching to the plugin. Then why not have the plugin vendor implement `oe_seal()` and `oe_unseal()` directly?

My another question is: how does that justify the `oe_seal_setting_*_t` structures? As I explained above, there are only 2 TEE neutral arguments - one selects KDK and the other supplies additional input to KDF. The rest are TEE specific and would be better not to be abstracted.

> By definition, the plugin already interacts with the OE SDK, and that doesn't change with the introduction of a formal plugin framework (otherwise, it's just a library and we wouldn't be having a conversation about adding/changing public OE SDK APIs at all). This simply formalizes the relationship between OE SDK and a plugin implementation, so that OE SDK can ensure a consistent user experience for how it interacts with the plugins on the user's behalf (including what it does and does not mediate).

Yes, "consistent user experience" is the key. But that doesn't mean we have to do it at the plugin level. I think "mediate" is the point. Could you provide an example of what to mediate by the top level APIs? If there's nothing to mediate, then there's no need for the top level API.

Plugin models work best in situations where the top level APIs need to translate a single invocation into potentially multiple interactions with a plugin, with a new set of parameters derived from the original ones passed into the top level. It doesn't look to me like the case here. 

Another good thing about plugins is they could be dynamically chosen at runtime. E.g., a TLS implementation could host multiple plugins, each for a different algorithm. Then 2 endpoints of a secure connection can negotiate on a commonly supported algorithm without users' intervention. But it isn't the case for sealing either. 

> Regarding whether it's simpler to link the plugin directly, the question is simpler for whom and in what context?
> - My priorities in thinking about the public API would be that the semantics for the OE SDK user are clear and easy to use, and I'm willing to trade off having the OE SDK & sealing implementation wrap a bit more work.

I'd argue the other way. Users would have to learn how to pass parameters in the form of `oe_seal_setting_*_t` structs. That's not incredibly complicated, but is still more complicated than just looking at API prototypes with self-explanatory arguments. It's more cumbersome to specify TEE specific parameters. Developers would have to study OE docs to figure out what settings correspond to which TEE params. They would have to guess the default values for those params if not specified explicitly. They would have to scratch their heads should they be unable to find an OE setting that corresponds to a param that needs to be set. But if we offer them the TEE interface struct (e.g., `KEYREQUEST` in SGX) directly with an initialization function (i.e. `oe_initialize_seal_key_info()`), all those problems will be gone.

> - In the case where we require a developer to link separate libraries and then invoke underlying methods provided by different libraries as needed, if you have more than one implementation, then the oe_seal() weak alias implementation is determined by link order and that can violate the Principle of Least Astonishment when a change in build definitions create a change in runtime behavior. It is not simpler in that regard.

Just to clarify, developers are not recommended to use `oe_seal()` if multiple implementations are linked. And if they have to do that (e.g., `oe_seal()` was referenced by a library whose source code isn't under the developer's control), the first implementation appearing on the linker command line is always the default. That's a well-defined behavior of `ld` and *weak symbols*. 

The plugin model also faces the same problem, as the developer has to have a means to specify the default plugin. The only difference is how. For weak symbols the order in which libs are supplied determines the default; while in the plugin models the source code (for registering plugins) would probably need changes. I don't see either is better than the other though.